Move Kostal Plenticore writable settings from sensor to select widget or switch

slowflyer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The "Battery Strategy" sensor has been removed from the Kostal Plenticore integration, it's replaced by a switch and a select entity to allow setting the battery charging settings from Home Assistant.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Depending on the needs writable sensors are replaced by a switch or select widget.

"Battery use strategy" is a switch now:
ON equals to "automatically" on the Plenticore battery settings page
OFF equals to "automatically economical" on the Plenticore battery settings page

The battery charging behaviour can be easily set by a select widget:
Battery:SmartBatteryControl:Enable equals to enabled "Activate smart battery control" on the Plenticore battery settings page
Battery:TimeControl:Enable equals to enabled "Time-controlled battery usage" on the Plenticore battery settings page
None equals to none of the above values selected on the Plenticore battery settings page

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

slowflyer committed:
Move "Battery:SmartBatteryControl:Enable" from a simple sensor to a switch
Add "Battery:TimeControl:Enable" as a switch

If you want to change charging behavior you need to turn off both switches, before you can enable the function you want. (Same as on Plenticore UI)

homeassistant commented:
Hi @slowflyer, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#56529).

Once you do that we will be able to review and accept this pull request.

Thanks!

emontnemery commented on homeassistant/components/kostal_plenticore/switch.py:
```diff
@@ -0,0 +1,173 @@
+        """Return true if device is on."""
+        return bool(self._state)
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique id of this switch Entity."""
+        return f"{self.entry_id}_{self.module_id}_{self.data_id}"
```
Set `_attr_unique_id` in the constructor instead, same for other constant properties

slowflyer replied:
Tried to set constant properties in the constructor

emontnemery commented on homeassistant/components/kostal_plenticore/switch.py:
```diff
@@ -0,0 +1,173 @@
+        """Return the device info."""
+        return self._device_info
+
+    @property
+    def assumed_state(self) -> bool:
+        """Return true if unable to access real state of entity."""
+        if self.coordinator.data is None:
+            raw_value = True
+            _LOGGER.debug("NO assumed_state ")
+        else:
+            raw_value = int(self.coordinator.data[self.module_id][self.data_id])
+
+        return bool(raw_value)
+
```
How is this helpful to the user?
Is it better to just mark the switch as unavailable?

slowflyer replied:
removed, was copied from an switchbot integration, does not make sense or does deliver valuable information

emontnemery commented:
> If you want to change charging behavior you need to turn off both switches, before you can enable the function you want. (Same as on Plenticore UI)

Does this mean it's three modes: No battery control, smart battery control, time control?
If so, instead of adding switches, a `select` entity is probably a better fit?

slowflyer committed:
removed:
    @property
    def assumed_state(self) -> bool

was copied from an switchbot integration, does not make sense or does deliver valuable information

Tried to set constant properties in the constructor

slowflyer commented:
> Does this mean there are three supported modes: No battery control, smart battery control, time control?
> If so, instead of adding switches, a `select` entity is probably a better fit?

At the end yes. You can have 

1. smart control
2. time usage control
3. no control

2. and 3. only makes a difference when you specify "Battery charge blocked, battery discharge permitted in the event of demand from house" and/or "Battery discharge blocked, battery charge permitted in the event of excess power" in a time table.

If you enable time usagee control without any restrictions, than 2. equals 3.

slowflyer committed:
correct typo, add new line at eof

slowflyer committed:
Initial state of switch was missing after (re)starting HA. Now working.

slowflyer committed:
Reformatted with black

emontnemery commented:
@slowflyer I suggested to model this as a select entity instead of as switch entities, if you agree on that please refactor this PR accordingly, if not, please explain why the implementation as switches is better.

slowflyer committed:
correct syntax errors from test run 09.10.2021

slowflyer commented:
@emontnemery: I already tried to explain that above: https://github.com/home-assistant/core/pull/56529#issuecomment-925142420
From my point of view it is most important to keep UI and functions comparable to the Plenticore UI. This will make it easier for users to understand, how things work and are related to each other.

slowflyer committed:
reformat

slowflyer committed:
update 15.10.2021

slowflyer commented:
> @slowflyer will you remove `switch.py`, I don't think it's needed anymore?

@emontnemery my idea was to remove  `switch.py` as soon as I have a working `select.py`. But currently I'm struggling to get that working.

slowflyer committed:
Set select value is working

slowflyer committed:
update 05.11.2021

slowflyer committed:
data correctly received

slowflyer committed:
working completly

slowflyer committed:
remove old switch definitions, now replaced by select widget

slowflyer commented:
At the end I got it work and now use a select instead of a switch to change between the 3 states for Battery charge:

1. smart control
2. time usage control
3. no control

the switch.py is still included, because it's a generic implementation which can be, and now is used, to switch the "Battery Strategy" between:

1. Automatic
2. Automatic economy

emontnemery commented on homeassistant/components/kostal_plenticore/const.py:
```diff
@@ -618,11 +618,25 @@
         "format_round",
     ),
+]
+SWITCH_SETTINGS_DATA = [
```
This is really quite unreadable, and the explanatory comment for `SENSOR_SETTINGS_DATA` doesn't seem to match `SWITCH_SETTINGS_DATA` and `SELECT_SETTINGS_DATA`.

Make these named tuples, dataclasses or maybe typed dicts to make them readable;

```
@dataclass
class SwitchSettings:
    ...

SWITCH_SETTINGS_DATA: list[SwitchSettings] = [
    ...
]
```

slowflyer replied:
Tried to solve that.

emontnemery replied:
Nice!

emontnemery commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -182,9 +187,137 @@ async def _async_update_data(self) -> dict[str, dict[str, str]]:
+            return True
+
+
+class PlenticoreSelectUpdateCoordinator(DataUpdateCoordinator):
```
What's the point of this class?

slowflyer replied:
PlenticoreSelectUpdateCoordinator was needed to pass the options for the select forward.
This is not possible with PlenticoreUpdateCoordinator.

emontnemery commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -182,9 +187,137 @@ async def _async_update_data(self) -> dict[str, dict[str, str]]:
+
+        return {module_id: {data_id: "None"}}
+
+    async def async_read_data(self, module_id: str, data_id: str) -> [str, bool]:
```
Why does this need to be overridden?

slowflyer replied:
Sorry, I do not get your point. What exactly do you mean?

emontnemery replied:
I mean that the implementation of `async_read_data` in `SelectDataUpdateCoordinator` is almost exactly the same as in `SettingDataUpdateCoordinator`. Can you try to make them share one implementation instead, maybe add the read/write in a mixin class?

slowflyer replied:
added a Mixin class

emontnemery commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -182,9 +187,137 @@ async def _async_update_data(self) -> dict[str, dict[str, str]]:
+        else:
+            return val
+
+    async def async_write_data(self, module_id: str, value: dict[str, str]) -> bool:
```
Why does this need to be overridden?

slowflyer replied:
Sorry, I do not get your point. What exactly do you mean?

emontnemery replied:
I mean that the implementation of `async_write_data` in `SelectDataUpdateCoordinator` is (now was) almost exactly the same as in `SettingDataUpdateCoordinator`. Can you try to make them share one implementation instead, maybe add the read/write in a mixin class?

slowflyer replied:
added a Mixin class

emontnemery commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -182,9 +187,137 @@ async def _async_update_data(self) -> dict[str, dict[str, str]]:
         fetched_data = await client.get_setting_values(self._fetch)
+        return fetched_data
+
+    async def async_read_data(self, module_id: str, data_id: str) -> [str, bool]:
```
This doesn't seem to be called from anywhere?

emontnemery commented on homeassistant/components/kostal_plenticore/select.py:
```diff
@@ -0,0 +1,137 @@
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+):
+    """Add kostal plenticore Sensors."""
```
stale

slowflyer replied:
Copied from senor.py to keep things consistent. Why was it ok, when maintainer submitted the code for senor.py?

emontnemery replied:
It was OK then because it was about sensors, now it's selects or switches.
Sorry for not being more clear about the comment.

slowflyer replied:
I still do not get the point. Looked at knx switch.py how they did it and rebuild it. If this is still "stale" I'm out of ideas ... if not, I will adopt switch.py in the same way.

emontnemery replied:
KNX just forgot to update their docstring, a PR fixing that would be welcome.
Anyhow, you fixed it in ca61960

emontnemery commented on homeassistant/components/kostal_plenticore/switch.py:
```diff
@@ -0,0 +1,183 @@
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+):
+    """Add kostal plenticore Sensors."""
```
stale

slowflyer replied:
Copied from senor.py to keep things consistent. Why was it ok, when maintainer submitted the code for senor.py?

emontnemery replied:
fixed in ca61960

slowflyer committed:
correct complaints from workflow run on 11/11/2021

slowflyer committed:
Add explanatory comment for switch and select

slowflyer committed:
Correct comments

slowflyer committed:
Removed function async def async_read_data(self, module_id: str, data_id: str)
from class SettingDataUpdateCoordinator

slowflyer committed:
Add Mixin class for read/write

slowflyer committed:
try to make select.py less "stale"

slowflyer commented:
@emontnemery may I ask you a question off-topic to this PR?

Why is this sensor not availbale on the energy dashboard?
```
    (
        "devices:local",
        "Grid_P",
        "Grid Power",
        {
            ATTR_UNIT_OF_MEASUREMENT: POWER_WATT,
            ATTR_DEVICE_CLASS: DEVICE_CLASS_POWER,
            ATTR_ENABLED_DEFAULT: True,
            ATTR_STATE_CLASS: STATE_CLASS_MEASUREMENT,
        },
        "format_round",
    ),
```

state_class and device_class are fine based on the documentation.

**Value entities - entities not representing a total amount​

Home Assistant tracks the min, max and mean value during the statistics period. The state_class property must be set to measurement, and the device_class must not be either of energy, gas, or monetary**

emontnemery commented:
@slowflyer I don't think the energy dashboard supports displaying power sensors. If the sensor settings are correct, it should be possible to add the power sensor to a lovelace statistics card however (https://www.home-assistant.io/lovelace/statistics-graph/)

Also off-topic:
Something is wrong with your development environment, failures from the linters (isort, black, etc.) should be caught by the pre-commit hooks.
Do you follow one of the suggested workflows here: https://developers.home-assistant.io/docs/development_environment ?

slowflyer committed:
new dev environment 2

homeassistant commented:
Hello @slowflyer,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`38aef7b9c4a0281435434d9aff912d1e4c44a950`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;56529&#x2F;commits&#x2F;38aef7b9c4a0281435434d9aff912d1e4c44a950) - This commit has something that looks like an email address (slowflyer@850168.no-reply.drupal.org). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

slowflyer committed:
new dev environment 2

slowflyer committed:
Merge branch 'create-writable-switches' of https://github.com/slowflyer/core into create-writable-switches

slowflyer committed:
correct syntax

slowflyer committed:
Merge branch 'home-assistant:dev' into create-writable-switches

slowflyer committed:
minor coding standard correction

slowflyer committed:
Merge branch 'create-writable-switches' of https://github.com/slowflyer/core into create-writable-switches

slowflyer commented:
> 
> 
> @slowflyer I don't think the energy dashboard supports displaying power sensors. If the sensor settings are correct, it should be possible to add the power sensor to a lovelace statistics card however (https://www.home-assistant.io/lovelace/statistics-graph/)
> 
> Also off-topic: Something is wrong with your development environment, failures from the linters (isort, black, etc.) should be caught by the pre-commit hooks. Do you follow one of the suggested workflows here: https://developers.home-assistant.io/docs/development_environment ?

I updated my dev-environment. Now I'm down to 1 error in hassfest:

  File "<unknown>", line 1
    ﻿"""Platform for Kostal Plenticore switches."""
    ^
SyntaxError: invalid character in identifier
Validating dependencies...
Error: Process completed with exit code 1.

No idea, what this message will tell me. I replaced this first line more than once - no change.

emontnemery commented:
hassfest does not like the UTF-8 BOM (https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8) which must have been inserted by whatever editor you're using.

Can you remove it please?
Do like this if you use vscode: https://stackoverflow.com/a/64194580

slowflyer committed:
Remove BOM

slowflyer committed:
Remove BOM on select.py

emontnemery commented:
Can you update the PR description please to explain what is now exposed as switch or select entities?

emontnemery commented:
Please update `.coveragerc` to exclude the new kostal plenticore files.

slowflyer committed:
Updated .coveragerc

emontnemery commented:
@slowflyer 
I updated the PR description to include a brief summary of the breaking change. Can you please review it?

slowflyer commented:
> @slowflyer I updated the PR description to include a brief summary of the breaking change. Can you please review it?

Thanks! Looks great!

emontnemery reviewed (approved):
Thanks, @slowflyer 👍

MartinHjelmare commented on homeassistant/components/kostal_plenticore/const.py:
```diff
@@ -688,11 +690,59 @@
+#  - on Label (str)
+#  - off Value (str)
+#  - off Label (str)
+SWITCH = namedtuple(
```
We prefer defining named tuples in the modern way as we do below for `SelectData` including type annotations.

slowflyer replied:
Fixed in https://github.com/home-assistant/core/pull/59998

MartinHjelmare commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -183,9 +222,83 @@ async def _async_update_data(self) -> dict[str, dict[str, str]]:
 
         return fetched_data
 
+    async def async_get_currentoption(
```
Please separate words with underscore in variable and function names.

slowflyer replied:
Fixed in https://github.com/home-assistant/core/pull/59998

MartinHjelmare commented on homeassistant/components/kostal_plenticore/select.py:
```diff
@@ -0,0 +1,129 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
```
Please type the whole signature when adding type annotations, ie type `async_add_entities`.

slowflyer replied:
Fixed in https://github.com/home-assistant/core/pull/59998

MartinHjelmare commented on homeassistant/components/kostal_plenticore/select.py:
```diff
@@ -0,0 +1,129 @@
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
```
Don't pass in `hass`. It will be set on the entity when the entity has been added to home assistant. This means you can access `hass` with `self.hass`, after the entity has been added to home assistant.

Pass in the coordinator instead.

slowflyer replied:
Fixed in https://github.com/home-assistant/core/pull/59998

MartinHjelmare commented on homeassistant/components/kostal_plenticore/switch.py:
```diff
@@ -0,0 +1,171 @@
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+        return {"last_run_success": self._last_run_success}
```
Why are we exposing this as a state attribute instead of raising an exception in the service handler?

slowflyer replied:
Fixed in https://github.com/home-assistant/core/pull/59998 by removing the property

MartinHjelmare commented on homeassistant/components/kostal_plenticore/select.py:
```diff
@@ -0,0 +1,129 @@
+        )
+
+        if is_available:
+            self._attr_current_option = self.coordinator.data[self.module_id][
```
Entity state properties shouldn't have side effects. We shouldn't change another attribute in the available property. Use a separate callback to update entity state if needed.

slowflyer replied:
Fixed in https://github.com/home-assistant/core/pull/59998

MartinHjelmare commented on homeassistant/components/kostal_plenticore/select.py:
```diff
@@ -0,0 +1,129 @@
+        self._attr_current_option = option
+        for all_option in self._attr_options:
+            if all_option != "None":
+                await self.coordinator.async_write_data(
```
Do we need to run these calls sequentially or can we gather the calls and run them concurrently?

slowflyer replied:
Yes we need, as the options in the select are 2 seperate entities in the Kostal Plenticore UI with the "disadvantage", that both options need to be disabled, before you can enable one.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

slowflyer commented on .coveragerc:
```diff
@@ -551,6 +551,8 @@ omit =
     homeassistant/components/kostal_plenticore/const.py
     homeassistant/components/kostal_plenticore/helper.py
     homeassistant/components/kostal_plenticore/sensor.py
+    homeassistant/components/kostal_plenticore/switch.py
```
Fixed in https://github.com/home-assistant/core/pull/59998

slowflyer commented on homeassistant/components/kostal_plenticore/__init__.py:
```diff
@@ -11,7 +11,7 @@
 _LOGGER = logging.getLogger(__name__)
 
-PLATFORMS = ["sensor"]
+PLATFORMS = ["sensor", "switch", "select"]
```
Fixed in https://github.com/home-assistant/core/pull/59998
